File: module-06/graphs/adventure.cpp
// adventure.cpp — CSC 249 Graph Homework Starter
// A navigable world built on a graph.
//
// Compile: g++ -std=c++17 adventure.cpp -o adventure
// Run:     ./adventure

#include <iostream>
#include <string>
#include <algorithm>
#include "Graph.h"

using namespace std;

// ──────────────────────────────────────────────────
// HELPER: Convert a string to lowercase
// ──────────────────────────────────────────────────
string toLower(string s) {
    transform(s.begin(), s.end(), s.begin(), ::tolower);
    return s;
}

// ──────────────────────────────────────────────────
// TODO: Build your world here
// ──────────────────────────────────────────────────
void buildWorld(Graph& world, Vertex*& startRoom) {
    // Example: a tiny 3-room house
    // Replace this with YOUR world (6+ rooms, 8+ connections)
    
    Vertex* kitchen = world.AddVertex("Kitchen");
    Vertex* hallway = world.AddVertex("Hallway");
    Vertex* bedroom = world.AddVertex("Bedroom");
    
    world.AddDirectedEdge(kitchen, hallway, "east");
    world.AddDirectedEdge(hallway, kitchen, "west");
    world.AddDirectedEdge(hallway, bedroom, "north");
    world.AddDirectedEdge(bedroom, hallway, "south");
    
    startRoom = kitchen;
}

// ──────────────────────────────────────────────────
// Display the current room and available exits
// ──────────────────────────────────────────────────
void look(Graph& world, Vertex* currentRoom) {
    cout << "\n────────────────────────" << endl;
    cout << "You are in the " << currentRoom->label << "." << endl;
    
    // TODO: Add room descriptions here (optional but encouraged)
    
    cout << "Exits:" << endl;
    for (Edge* exit : *world.GetEdgesFrom(currentRoom)) {
        cout << "  " << exit->weight << " -> " << exit->toVertex->label << endl;
    }
    cout << "────────────────────────" << endl;
}

// ──────────────────────────────────────────────────
// Try to move in a direction. Returns new room or nullptr.
// ──────────────────────────────────────────────────
Vertex* tryMove(Graph& world, Vertex* currentRoom, const string& direction) {
    for (Edge* exit : *world.GetEdgesFrom(currentRoom)) {
        if (toLower(exit->weight) == toLower(direction)) {
            return exit->toVertex;
        }
    }
    return nullptr;  // no exit in that direction
}

// ──────────────────────────────────────────────────
// Main game loop
// ──────────────────────────────────────────────────
int main() {
    Graph world;
    Vertex* currentRoom = nullptr;
    
    buildWorld(world, currentRoom);
    
    cout << "========================================" << endl;
    cout << "  Welcome to [YOUR WORLD NAME HERE]" << endl;
    cout << "  Type a direction to move, or:" << endl;
    cout << "    look  - see where you are" << endl;
    cout << "    exits - list available exits" << endl;
    cout << "    quit  - exit the game" << endl;
    cout << "========================================" << endl;
    
    look(world, currentRoom);
    
    string input;
    while (true) {
        cout << "\n> ";
        getline(cin, input);
        input = toLower(input);
        
        if (input == "quit" || input == "q") {
            cout << "Thanks for exploring!" << endl;
            break;
        }
        else if (input == "look" || input == "l") {
            look(world, currentRoom);
        }
        else if (input == "exits") {
            for (Edge* exit : *world.GetEdgesFrom(currentRoom)) {
                cout << "  " << exit->weight << endl;
            }
        }
        else if (input.substr(0, 3) == "go ") {
            // Handle "go north" style input
            string direction = input.substr(3);
            Vertex* nextRoom = tryMove(world, currentRoom, direction);
            if (nextRoom) {
                currentRoom = nextRoom;
                look(world, currentRoom);
            } else {
                cout << "You can't go that way." << endl;
            }
        }
        else {
            // Try the raw input as a direction (e.g., just "north")
            Vertex* nextRoom = tryMove(world, currentRoom, input);
            if (nextRoom) {
                currentRoom = nextRoom;
                look(world, currentRoom);
            } else {
                cout << "I don't understand '" << input << "'." << endl;
                cout << "Try a direction, 'look', 'exits', or 'quit'." << endl;
            }
        }
    }
    
    return 0;
}

File: module-06/graphs/adventure.py
# adventure.py — CSC 249 Graph Homework Starter
# A navigable world built on a graph.
#
# Run: python adventure.py

from graph import Graph


def build_world(world: Graph) -> "Vertex":
    """
    Build the rooms and connections for your world.
    Returns the starting room vertex.
    
    TODO: Replace this example with YOUR world (6+ rooms, 8+ connections)
    """
    # Example: a tiny 3-room house
    kitchen = world.add_vertex("Kitchen")
    hallway = world.add_vertex("Hallway")
    bedroom = world.add_vertex("Bedroom")
    
    world.add_directed_edge(kitchen, hallway, "east")
    world.add_directed_edge(hallway, kitchen, "west")
    world.add_directed_edge(hallway, bedroom, "north")
    world.add_directed_edge(bedroom, hallway, "south")
    
    return kitchen  # starting room


def look(world: Graph, current_room) -> None:
    """Display the current room and available exits."""
    print()
    print("────────────────────────")
    print(f"You are in the {current_room.label}.")
    
    # TODO: Add room descriptions here (optional but encouraged)
    
    print("Exits:")
    for exit_edge in world.get_edges_from(current_room):
        print(f"  {exit_edge.weight} -> {exit_edge.to_vertex.label}")
    print("────────────────────────")


def try_move(world: Graph, current_room, direction: str):
    """Try to move in a direction. Returns new room or None."""
    for exit_edge in world.get_edges_from(current_room):
        if exit_edge.weight.lower() == direction.lower():
            return exit_edge.to_vertex
    return None


def main():
    world = Graph()
    current_room = build_world(world)
    
    print("========================================")
    print("  Welcome to [YOUR WORLD NAME HERE]")
    print("  Type a direction to move, or:")
    print("    look  - see where you are")
    print("    exits - list available exits")
    print("    quit  - exit the game")
    print("========================================")
    
    look(world, current_room)
    
    while True:
        user_input = input("\n> ").strip().lower()
        
        if user_input in ("quit", "q"):
            print("Thanks for exploring!")
            break
        
        elif user_input in ("look", "l"):
            look(world, current_room)
        
        elif user_input == "exits":
            for exit_edge in world.get_edges_from(current_room):
                print(f"  {exit_edge.weight}")
        
        elif user_input.startswith("go "):
            direction = user_input[3:]
            next_room = try_move(world, current_room, direction)
            if next_room:
                current_room = next_room
                look(world, current_room)
            else:
                print("You can't go that way.")
        
        else:
            # Try raw input as a direction (e.g., just "north")
            next_room = try_move(world, current_room, user_input)
            if next_room:
                current_room = next_room
                look(world, current_room)
            else:
                print(f"I don't understand '{user_input}'.")
                print("Try a direction, 'look', 'exits', or 'quit'.")


if __name__ == "__main__":
    main()
